lib/timers.js

```diff
@@ -483,152 +575,3 @@ exports.clearImmediate = function(immediate) {
-  if (msecs <= 0)
-    return;
-
-  if (!timer._onTimeout)
-    return;
-
-  if (domain) {
-    if (domain._disposed)
-      return;
-
-    domain.enter();
-  }
-
-  debug('unreftimer firing timeout');
-  timer._called = true;
-  _runOnTimeout(timer);
-
-  if (domain)
-    domain.exit();
-}
-
-function _runOnTimeout(timer) {
-  var threw = true;
-  try {
-    timer._onTimeout();
-    threw = false;
-  } finally {
-    if (threw) process.nextTick(unrefTimeout);
-  }
-}
-
-function unrefTimeout() {
-  var now = Timer.now();
-
-  debug('unrefTimer fired');
-
-  var timeSinceLastActive;
-  var nextTimeoutTime;
-  var nextTimeoutDuration;
-  var minNextTimeoutTime = TIMEOUT_MAX;
-  var timersToTimeout = [];
-
-  // The actual timer fired and has not yet been rearmed,
-  // let's consider its next firing time is invalid for now.
-  // It may be set to a relevant time in the future once
-  // we scanned through the whole list of timeouts and if
-  // we find a timeout that needs to expire.
-  unrefTimer.when = -1;
-
-  // Iterate over the list of timeouts,
-  // call the onTimeout callback for those expired,
-  // and rearm the actual timer if the next timeout to expire
-  // will expire before the current actual timer.
-  var cur = unrefList._idlePrev;
-  while (cur !== unrefList) {
-    timeSinceLastActive = now - cur._idleStart;
-
-    if (timeSinceLastActive < cur._idleTimeout) {
-      // This timer hasn't expired yet, but check if its expiring time is
-      // earlier than the actual timer's expiring time
-
-      nextTimeoutDuration = cur._idleTimeout - timeSinceLastActive;
-      nextTimeoutTime = now + nextTimeoutDuration;
-      if (minNextTimeoutTime === TIMEOUT_MAX ||
-          (nextTimeoutTime < minNextTimeoutTime)) {
-        // We found a timeout that will expire earlier,
-        // store its next timeout time now so that we
-        // can rearm the actual timer accordingly when
-        // we scanned through the whole list.
-        minNextTimeoutTime = nextTimeoutTime;
-      }
-    } else {
-      // We found a timer that expired. Do not call its _onTimeout callback
-      // right now, as it could mutate any item of the list (including itself).
-      // Instead, add it to another list that will be processed once the list
-      // of current timers has been fully traversed.
-      timersToTimeout.push(cur);
-    }
-
-    cur = cur._idlePrev;
-  }
-
-  var nbTimersToTimeout = timersToTimeout.length;
-  for (var timerIdx = 0; timerIdx < nbTimersToTimeout; ++timerIdx)
-    _makeTimerTimeout(timersToTimeout[timerIdx]);
-
-
-  // Rearm the actual timer with the timeout delay
-  // of the earliest timeout found.
-  if (minNextTimeoutTime !== TIMEOUT_MAX) {
-    unrefTimer.start(minNextTimeoutTime - now, 0);
-    unrefTimer.when = minNextTimeoutTime;
-    debug('unrefTimer rescheduled');
-  } else if (L.isEmpty(unrefList)) {
-    debug('unrefList is empty');
-  }
-}
-
-
-exports._unrefActive = function(item) {
```

Comment:
> I believe this was done to prevent the creation of a lot of underlying timers and because at that time the implementers believed that this would hurt performance.

> Does anyone have more context around this (benchmarks, sample code that shows pathological cases, etc.)? I started working on timers after internal timers were implemented, and so I'm not familiar with the details behind that decision.

Nothing personal to TJ, but due to the fact that the original internal timers implementation was a sorted linked-list insert (`O(n)`), I'm lead to believe that TJ quite possibly just didn't fully understand the regular implementation. Considering there was no review it's quite probable the current internal timers implementation was somewhat of an impulsive mistake. https://github.com/nodejs/node/commit/f46ad012bc5a40194242ea1e9669c9cc25bd7047

This actually reverts the timer implementation to the original, with the added benefit that it also supports implicitly unreferenced (internal) timers. There were never perf problems with the original as far as I can tell.